server/neptune/workflows/internal/terraform/workflow.go

```diff
@@ -47,37 +49,51 @@ type workerActivities interface {
 	TerraformInit(context.Context, activities.TerraformInitRequest) error
 	TerraformPlan(context.Context, activities.TerraformPlanRequest) error
 	TerraformApply(context.Context, activities.TerraformApplyRequest) error
-	ExecuteCommand(context.Context, activities.ExecuteCommandRequest) error
+	ExecuteCommand(context.Context, activities.ExecuteCommandRequest) activities.ExecuteCommandResponse
 	Notify(context.Context, activities.NotifyRequest) error
 	Cleanup(context.Context, activities.CleanupRequest) error
 }
 
 type Runner struct {
 	Activities workerActivities
+	JobRunner  runners.JobRunner
 	Request    Request
 }
 
 func newRunner(ctx workflow.Context, request Request) *Runner {
 	var a *activities.Terraform
+
+	runStepRunner := &runners.RunStepRunner{
+		Activity: a,
+	}
 	return &Runner{
 		Activities: a,
 		Request:    request,
+		JobRunner: runners.NewJobRunner(
+			runStepRunner,
+			&runners.EnvStepRunner{
+				RunStepRunner: runStepRunner,
+			},
+		),
 	}
 }
 
 func (r *Runner) Run(ctx workflow.Context) error {
+	// Root instance has all the metadata needed to execute a step in a root
+	rootInstance := steps.BuildRootInstanceFrom(r.Request.Root, r.Request.Repo)
+	executionContext := steps.BuildExecutionContextFrom(ctx, r.Request.Root, &map[string]string{})
+
 	// Clone repository into disk
 	err := workflow.ExecuteActivity(ctx, r.Activities.GithubRepoClone, activities.GithubRepoCloneRequest{}).Get(ctx, nil)
 	if err != nil {
 		return errors.Wrap(err, "executing GH repo clone")
 	}
-	// Run plan steps
-	for _, step := range r.Request.Root.Plan.Steps {
-		err := r.runStep(ctx, step)
-		if err != nil {
-			return errors.Wrap(err, "running step")
-		}
+
+	_, err = r.JobRunner.Run(*executionContext, r.Request.Root.Plan, rootInstance)
```

Comment:
`steps.BuildExecutionContext` should just return a pointer, you shouldn't reference it with each call.